partiql-planner/src/main/kotlin/org/partiql/planner/typer/PlanTyper.kt

```diff
@@ -531,24 +542,94 @@ internal class PlanTyper(
             }
         }
 
-        override fun visitRexOpCase(node: Rex.Op.Case, ctx: StaticType?): Rex {
-            val visitedBranches = node.branches.map { visitRexOpCaseBranch(it, null) }
+        override fun visitRexOpCase(node: Rex.Op.Case, ctx: StaticType?): Rex = plan {
+            // Type and simplify branches
+            val visitedBranches = node.branches.map { visitRexOpCaseBranch(it, it.rex.type) }
+
+            // Prune branches known to never execute
+            val newBranches = visitedBranches.mapNotNull { branch ->
+                val conditionFoldedOp = branch.condition.op as? Rex.Op.Lit ?: return@mapNotNull branch
+                val conditionBooleanLiteral = conditionFoldedOp.value as? BoolValue ?: return@mapNotNull branch
+                when (conditionBooleanLiteral.value) {
+                    false -> null
+                    else -> branch
+                }
+            }
+
+            // Calculate final expression (short-circuit to first branch if the condition is always TRUE).
             val resultTypes = visitedBranches.map { it.rex }.map { it.type }
-            return rex(AnyOfType(resultTypes.toSet()).flatten(), node.copy(branches = visitedBranches))
+            val firstBranch = newBranches.firstOrNull() ?: error("CASE_WHEN has NO branches.")
+            return@plan when (isLiteralTrue(firstBranch.condition)) {
+                true -> firstBranch.rex
+                false -> rex(type = StaticType.unionOf(resultTypes.toSet()).flatten(), node.copy(branches = newBranches))
+            }
```

Comment:
This is entirely stylistic, but I'm leaving this here for some info / discussion. I've found loop control flow when using map / fold / mapNotNull to be more difficult to reason about than just using mutable lists and simple for loops; when I believe the stdlib code implements these functions with mutable lists. In this case, we don't actually map the elements so a simple `filter` would suffice. Consider the snippet

```kotlin
val newBranches = visitedBranches.mapNotNull { branch ->
    val conditionFoldedOp = branch.condition.op as? Rex.Op.Lit ?: return@mapNotNull branch
    val conditionBooleanLiteral = conditionFoldedOp.value as? BoolValue ?: return@mapNotNull branch
    when (conditionBooleanLiteral.value) {
        false -> null
        else -> branch
     }
}

// compared to
val newBranches = visitedBranches.filterNot { isLiteralBool(it.condition, false) }
```

Reply:
Now that I'm looking at it you can reduce a lot. You also don't need the plan block.

```kotlin
val branches = node.branches
                   .map { visitRexOpCaseBranch(it, it.rex.type) }
                   .filterNot { isLiteralBool(it.condition, false) }
return when (branches.size) {
    0 -> error("CASE_WHEN has NO branches.")
    1 -> branches.first().rex
    else -> {
         val types = branches.map { it.rex.type }.toSet()
         val type = AnyOfType(types).flatten()
         val op = rexOpCase(branches)
         rex(type, op)
    }
}
```

Reply:
Agreed. The logic is a bit different, but yeah I'll update this. Good call.